dotcom-rendering/src/web/components/PinnedPostLiveness.importable.tsx

```diff
@@ -25,10 +27,33 @@ export const PinnedPostLiveness = ({}) => {
 		pinnedPost && pinnedPost.scrollHeight <= pinnedPost.clientHeight;
 	if (contentFitsContainer) hideShowMore();
 
+	const hasBeenSeen = useRef(false);
```

Comment:
Why are we using a ref here? So it doesn’t reset between renders?

Reply:
This can be fixed forward as far as I’m concerned. Not a blocker. If we can use a state directly, though, we probably should, as this is the main benefit of (P)React!

Reply:
Given that this only gets set once, could we use a let here instead of useRef?

Reply:
You’d probably want to `useState`, but I have to say I can’t quite follow timing logic, so if this works, I’m happy.

If we want to keep things as functional as possible, there is this library which provides the observer as a hook: https://github.com/asyarb/use-intersection-observer

```ts
const [hasBeenSeen, setHasBeenSeen] = useState<boolean>(false);

const isInView = useIntersectionObserver({ element: pinnedPost, threshold: 0.1 })

useEffect(() => {
  if(!hasBeenSeen && isInView) {
    setHasBeenSeen(true);
    pinnedPostTiming.clear(); // I think this will no longer be necessary
    pinnedPostTiming.start()
  } else if (hasBeenSeen) {
    pinnedPostTiming.end()
  }
}, [hasBeenSeen, isInView])

```

Reply:
Thanks Max! That's really interesting. I'll leave it as is for now and merge so as not to block but maybe I'll put a new ticket around this as a possible refactor.
I'll also tag @marjisound here too as I think she'll find this interesting.

Reply:
We already have a custom hook for the intersection observer in DCR [here](https://github.com/guardian/dotcom-rendering/blob/02177fb7f911eb5e2ba50c99ebd4a76342a61bd6/dotcom-rendering/src/web/lib/useIsInView.ts)

Reply:
Hi @mxdvl was just looking at the suggestion you made and had a chat with @oliverlloyd too. I tried using the `useIsInView` hook but it doesn't work in this instance. That's because  `setNode` expects the element it is being used on to be rendered by React. And in our case the element is a dom element we retrieved through query selector `window.document.querySelector('[data-gu-marker=pinned-post]')`

So when I try `setNode(pinnedPost);` I get the following error: 
![image](https://user-images.githubusercontent.com/15894063/162188621-68baacea-f8ef-4b4d-8e27-d8d63f4ddf05.png)

Regarding the other suggestion you had made for using `useState` instead of `useRef` for `hasBeenSeen`: 

I went with the `useRef` approach because I thought it's not necessary to have a re-render when the `hasBeenSeen` value changes. Although, the `hasBeenSeen` is supposed to be changed only the first time intersection happens. And after then the whole time the user stays in the page, the `hasBeenSeen` doesn't change at all. So if we use `useState` we'd only have 1 extra re-render. 

The current work using the `useRef` works fine, but do you think using the `useState` will bring any advantages? I'm happy to change it if it'll give us any benefit :)

Reply:
The type error you’re getting is only a compiler error, which you can circumvent by telling TS that you’re selecting an `HTMLElement`, not just any `Element`:

```ts
const pinnedPost: HTMLElement | null = !isServer
	? window.document.querySelector('[data-gu-marker=pinned-post]')
	: null;
const pinnedPostCheckBox: HTMLElement | null = !isServer
	? window.document.querySelector('input[name=pinned-post-checkbox]')
	: null;
```

Regarding `useRef` v `useState`, I don’t think we’d have a way to measure the performance impact of choosing one over the other anyways. Let’s keep it as-is.